File: UploadClient/Protocol.cs
using System.Text;

namespace UploadClient;


/// <summary>
/// Byte ordering
/// </summary>
enum PB : byte
{
    ENQ = 0x05, // Anfrage/Start,
    ETX = 0x03, // Ende der Nachricht
    US = 0x1F, // Einheitentrenner/Blocktrenner
}

/// <summary>
/// Protocol operation
/// </summary>
enum POP : byte
{
    UPLOAD,
    DOWNLOAD
}

// ENQ operation *length+header US body ETX 
struct PHeader
{
    public string Key { get; init; }

    public string Content { get; init; }

    public PHeader(string key, string content)
    {
        Key = key;
        Content = content;
    }

    public void Write(Stream stream)
    {
        byte[] keyB = Encoding.UTF8.GetBytes(Key); 
        byte[] valueB = Encoding.UTF8.GetBytes(Content);

        int length = keyB.Length + 1 + valueB.Length;

        stream.Write(BitConverter.GetBytes(length));
        stream.Write(keyB);
        stream.WriteByte((byte)'=');
        stream.Write(valueB);
    }

    public static PHeader Read(Stream stream)
    {
        byte[] lenBytes = new byte[4];
        stream.ReadExactly(lenBytes);

        int length = BitConverter.ToInt32(lenBytes);

        byte[] buffer = new byte[length];
        stream.ReadExactly(buffer);

        string content = Encoding.UTF8.GetString(buffer);
        string[] split = content.Split("=");

        if (split.Length != 2)
        {
            throw new Exception("Invalid header!");
        }

        return new PHeader(split[0], split[1]);
    }
}

internal class PMessage
{
    public POP Operation { get; init; }

    public List<PHeader> Headers { get; init; }

    public string Body { get; set; }

    public PMessage(POP operation, string body)
        : this(operation, Enumerable.Empty<PHeader>(), body)
    {

    }

    public PMessage(
        POP operation, 
        IEnumerable<PHeader> headers, 
        string body)
    {
        Operation = operation;
        Headers = new List<PHeader>(headers);
        Body = body;
    }

    public void Write(Stream stream)
    {
        stream.WriteByte((byte)PB.ENQ);
        stream.WriteByte((byte)Operation);

        foreach (PHeader header in Headers)
        {
            header.Write(stream);
        }

        stream.WriteByte((byte)PB.US);
        byte[] body = Encoding.UTF8.GetBytes(Body);
        stream.Write(body);

        stream.WriteByte((byte)PB.ETX);

        stream.Flush();
    }

    public static PMessage Read(Stream stream)
    {
        PB start = (PB)stream.ReadByte();

        if (start != PB.ENQ)
        {
            throw new Exception("Not a protocol message!");
        }

        POP operation = (POP)stream.ReadByte();
        if (!Enum.IsDefined(typeof(POP), operation))
        {
            throw new Exception("Not a valid operation");
        }

        while(true)
        {
            PHeader pheader = 
        }
    }
}

File: upload_server/src/protocol.rs
use std::io::{Write, Read};

#[derive(Debug, Copy, Clone)]
pub enum PB {
    ENQ = 0x05, // Anfrage/Start
    RS = 0x1E, // Header start
    US = 0x1F
}

#[derive(Debug, Copy, Clone)]
pub enum POP {
    UPLOAD = 0x00,
    DOWNLOAD = 0x01
}

pub struct PHeader {
    key: String, 
    value: String
}

impl PHeader {
    pub fn new(key: String, value: String) -> Self {
        PHeader {
            key: key,
            value: value,
        }
    }

    pub fn read(stream: impl Read) -> Self {
        Self {
            key: String::from("Hello"),
            value: String::from("world"),
        }
    }

    pub fn write(&self, stream: &mut impl Write) {
        let key_b = self.key.as_bytes();
        let mut value_b = self.value.as_bytes();

        let len: i32 = i32::try_from(key_b.len() + 1 + value_b.len()).unwrap();

        let len_b = len.to_ne_bytes();
        stream.write_all(&len_b);
        stream.write_all(&key_b);
        
        let mut last = Vec::from("=".as_bytes());
        let mut tmp = Vec::from(value_b);
        last.append(&mut tmp);

        stream.write_all(&tmp);
    }

}

pub struct PMessage {
    operation: POP,
    headers: Vec<PHeader>,
    body: String
}

impl PMessage {
    pub fn new(op: POP, h: Vec<PHeader>, b: String) -> Self {
        PMessage {
            operation: op,
            headers: h,
            body: b,
        }
    }

    pub fn read(stream: impl Read) -> Self {
        todo!();
    }

    pub fn write(&self, stream: &mut impl Write) {

        let start: [u8; 2] = [PB::ENQ as u8, self.operation as u8];
        stream.write_all(&start);

        for header in self.headers.iter().enumerate() {
            let hs: [u8; 1] = [PB::RS as u8];
            stream.write_all(&hs);
            header.1.write(stream);
        }

        let mut end: [u8; 1] = [PB::US as u8];

        let body = self.body.as_bytes();
        let len: i32 = i32::try_from(body.len()).unwrap();

        let len_b = len.to_ne_bytes();
        
        stream.write(&end);
        stream.write_all(&len_b);
        stream.write_all(&body);

        stream.flush();
    }
}